### ISSUE TITLE
MEMLEAK with SMS_Ld20.f45_f45.IELMFATES.xxx.elm-fates_eca on KNL and perlmutter

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#4709]
There are 2 tests that are failing with MEMLEAK on the 2 KNL machines (cori/theta):

```
SMS_Ld20.f45_f45.IELMFATES.cori-knl_intel.elm-fates_eca
SMS_Ld20.f45_f45.IELMFATES.cori-knl_intel.elm-fates_rd
```

I tried `SMS_Lm3.f45_f45.IELMFATES.cori-knl_intel.elm-fates_eca` and see the memory continue to grow.
I also tried GNU compiler (which is v8 on cori) and see the same issue.
Building DEBUG also shows the same type of memory growth (for both Intel/GNU).

The test passes using cori-haswell (with Intel and GNU) and presumably on other machines as I'm not seeing this test report a failure.

Although, I do see the same fail with those 2 tests on Perlmutter which is using GNU v9 on AMD's.

**Comment by ndkeen** [comment]
I tried `azamat/elm/fix-datatype-memleaks` and I do see a decrease in memory use, but it's still clearly growing -- will include a plot when I get a chance to make one.

I do see that `SMS_Ld20.f45_f45.IELMFATES.cori-haswell_intel.elm-fates_rd` jsut failed with MEMLEAK on both cori-haswell and anvil -- so maybe it really is a memory leak on any machine, but the amount measured is near the threshold for causing the test to fail.

**Comment by rgknox** [comment]
Looking into this too

**Comment by rgknox** [comment]
Is it possible this is a red herring, false positive?  As a typical simulation progresses, the FATES model will continue to grow in memory as its representation of biodiversity increases, by design.  FATES is constantly spawning new objects (plant cohorts and area patches) and cleaning up old.  At the beginning of a cold-started simulation this is particularly unbalanced, there is more spawning of new objects  than there are cleaning up of old.  FATES will eventually hit a memory usage equilibrium of sorts, but that will not be achieved in these tests.

**Comment by ndkeen** [comment]
Every time I run the problem longer, the memory continues to increase. However, it's true that while memory increase may look odd, it's not a terrible amount and the rate of increase slows.

Here is one attempt at running several years (only completed about 5 years).
![memhighwater](https://user-images.githubusercontent.com/5684727/148838362-ec56017a-d9d5-478d-916d-7248f00da96d.png)

And as I thought it was interesting, the wall-clock time per day also increased.. but not clear what's happening there.

![fates-time](https://user-images.githubusercontent.com/5684727/148838439-cbfe0540-03d1-4096-8885-be0e36c7f978.png)

**Comment by ndkeen** [comment]
Actually on cori-knl, when I try to run the problem longer, I hit a failure. It happened twice.  Again this is with the branch noted above. This is with `SMS_Ly1.f45_f45.IELMFATES.cori-knl_intel.elm-fates_eca`

```
33:  column nbalance error =   3.326669506359338E-003        9134        5282
33:  Latdeg,Londeg         =    10.0000000000000        105.000000000000     
33:  begnb                 =    1.97744198998876     
33:  endnb                 =    1.97415544228068     
33:  delta store           =  -3.286547708081677E-003
33:  input mass            =   4.012197523323643E-005
33:  output mass           =   1.769555761410518E-010
33:  net flux              =   4.012179827766029E-005
33:  denit                 =   0.000000000000000E+000
33:  n2onit                =   1.769556022419521E-010
33:  no3 leach             =   0.000000000000000E+000
33:  no3 runof             =   0.000000000000000E+000
33:  ndep                  =   3.493213860714036E-005
33:  nfix                  =   5.141157170586869E-006
33:  nsup                  =   0.000000000000000E+000
33:  n_to_plant            =   0.000000000000000E+000
33:  plant_to_litter       =   4.867945550919808E-008
33:  fire                  =   0.000000000000000E+000
33:  ENDRUN:
33:  ERROR in EcosystemBalanceCheckMod.F90 at line 541                              
33:                                                                                 
33:                                                                                 
33:                                                                                 
33:                                                                                 
33:                                                                                 
33:                                        
33:  ERROR: Unknown error submitted to shr_abort_abort.
33: Image              PC                Routine            Line        Source             
33: e3sm.exe           0000000002DD0AA5  Unknown               Unknown  Unknown
33: e3sm.exe           00000000012A9231  shr_abort_mod_mp_         114  shr_abort_mod.F90
33: e3sm.exe           000000000051B77F  abortutils_mp_end          43  abortutils.F90
33: e3sm.exe           0000000000C61447  ecosystembalancec         541  EcosystemBalanceCheckMod.F90
33: e3sm.exe           000000000052E1DC  elm_driver_mp_elm        1293  elm_driver.F90
33: e3sm.exe           000000000050E54D  lnd_comp_mct_mp_l         506  lnd_comp_mct.F90
33: e3sm.exe           0000000000428029  component_mod_mp_         751  component_mod.F90
33: e3sm.exe           0000000000404A80  cime_comp_mod_mp_        2876  cime_comp_mod.F90
33: e3sm.exe           0000000000427C37  MAIN__                    153  cime_driver.F90
33: e3sm.exe           0000000000401F72  Unknown               Unknown  Unknown
33: e3sm.exe           0000000002F6E10F  Unknown               Unknown  Unknown
33: e3sm.exe           0000000000401E5A  Unknown               Unknown  Unknown
33: Rank 33 [Fri Jan  7 20:18:18 2022] [c1-1c0s5n0] application called MPI_Abort(MPI_COMM_WORLD, 1001) - process 33

case:
/global/cscratch1/sd/ndk/e3sm_scratch/cori-knl/azelm/SMS_Ly1.f45_f45.IELMFATES.cori-knl_intel.elm-fates_eca.20220107_110638_i6qgim
```

**Comment by rgknox** [comment]
@ndkeen those are some interesting plots, thanks.  The results that you show (above before reporting the crash) are what I would expect.  It will take on the order of years, if not decades, for FATES memory usage to hit an equilibrium.  As the simulation progresses, plant biodiversity increases as plants of different types grow to different sizes.  To accomodate this, FATES allocates new objects to handle these the plant groups.  This is core to the model's functionality.   Like with memory, with more plants represented as objects, the physics routines then run over larger lists of objects, thus the simulations take more time.   It is interesting that the wallclock time does not show the same patterning as the memory usage.  I guess the annual cycle obscures it...?

**Comment by rgknox** [comment]
Regarding this crash... We've been working on an updated version of the FATES-ELM nutrient coupling, which this test covers. Most of my testing on has been at a single site, but I can try a long term global smoke test. I can certainly try a long-term smoke test on cori KNL before I submit the PR.  If anyone is curious, here are the branches:
https://github.com/rgknox/fates/tree/cnp-dynamic-root-allom
https://github.com/rgknox/E3SM/tree/rgknox/lnd%2Fupdate-suppl-fates-uptake-v2

**Comment by ckoven** [comment]
Yes, thanks @ndkeen for these plots.  Just to second @rgknox's point, if it helps to visualize why the memory structures take so long to stabilize, the link below is to an animation that I generated using annual snapshots of FATES restart files, of 600 years of forest dynamics at a single site, which I posted as a video supplement to our 2020 paper in Biogeosciences.  Each of the popsicle-shaped things is a FATES cohort of trees, and the vertical dashed lines separate the patches of ground that the cohorts of trees grow on (arranged from the most-recently-disturbed ground on the left to the oldest-time-since-disturbed ground on the right).  The dynamic heterogeneity of the forest doesn't stabilize until about a third of the way (so ~200 years) into the simulation, and so the memory structures will tend to continue to grow until it finds that stability.

https://doi.org/10.5446/43627
